### PR TITLE
[libunwind][AIX] Remove weak declaration "__xlcxx_personality_v0"

### PR DESCRIPTION
**Pull request by xingxue-ibm**
`__xlcxx_personality_v0` is the personality routine in `libc++abi` for the EH of applications generated by the legacy IBM C++ compiler. Since the EH info generated by the legacy compiler does not provide the location of the personality routine, this routine is hard-coded as the handler for legacy EH in the unwinder. The symbol is resolved dynamically using `dlopen()` to avoid a hard dependency of `libunwind` on `libc++abi` for cases such as non-C++ applications. The weak declaration of `__xlcxx_personality_v0` was originally intended to bypass `dlopen()` if the C++ application generated by the legacy compiler is statically linked with the new LLVM C++ compiler. Unfortunately, this causes problems with runtime linking for Clang-compiled code using the unwinder that does not link with `libc++abi`.

On the other hand, the C++ runtime libraries shipped for AIX are actually stripped and statically linking is not supported. So, we can fix the problem by removing the `__xlcxx_personality_v0` weak declaration. Besides, `dlopen()` would work as long as the libc++abi shared library is available.

### CODE REVIEW

**Review by hubert-reinterpretcast — COMMENTED** [review]
I think a small test case could be added with a `int main(void) {}` C program linked with -`brtl` and `LDR_PRELOAD`/`LDR_PRELOAD64` set to `libunwind.a(libunwind.so.1)`.

**Comment by hubert-reinterpretcast** [comment]
@xingxue-ibm, I edited the PR description text. Please check my edits.

**Comment by xingxue-ibm** [comment]
> @xingxue-ibm, I edited the PR description text. Please check my edits.
Thanks for the editing, @hubert-reinterpretcast!

File: libunwind/src/UnwindCursor.hpp [hunk 1]
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
...
     // table and is exported from libc++abi, is directly assigned as the
     // handler here. When a legacy XLC++ frame is encountered, the symbol
-    // is resolved dynamically using dlopen() to avoid hard dependency from
-    // libunwind on libc++abi.
+    // is resolved dynamically using dlopen() to avoid a hard dependency of
+    // libunwind on libc++abi in cases such as non-C++ applications.
 
     // Resolve the function pointer to the state table personality if it has
-    // not already.
+    // not already done.
```
[the remark this answers is not in the document]
Comment (xingxue-ibm) [reply libunwind/src/UnwindCursor.hpp]: Fixed, thanks!

File: libunwind/src/UnwindCursor.hpp [hunk 2]
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
...
-                                       errno);
-            assert(0 && "dlopen() failed");
-          }
-          xlcPersonalityV0 = reinterpret_cast<__xlcxx_personality_v0_t *>(
-              dlsym(libHandle, "__xlcxx_personality_v0"));
-          if (xlcPersonalityV0 == NULL) {
-            _LIBUNWIND_TRACE_UNWINDING("dlsym() failed with errno=%d\n", errno);
-            assert(0 && "dlsym() failed");
-          }
-          dlclose(libHandle);
```
[the remark this answers is not in the document]
Comment (xingxue-ibm) [reply libunwind/src/UnwindCursor.hpp:2162]: Restored the `dlclose` call. Will do it in a separate PR.

File: libunwind/src/UnwindCursor.hpp [hunk 3]
```diff
@@ -2126,42 +2125,34 @@ bool UnwindCursor<A, R>::getInfoFromTBTable(pint_t pc, R &registers) {
...
-    // not already.
+    // not already done.
     if (xlcPersonalityV0 == NULL) {
       xlcPersonalityV0InitLock.lock();
       if (xlcPersonalityV0 == NULL) {
-        // If libc++abi is statically linked in, symbol __xlcxx_personality_v0
-        // has been resolved at the link time.
-        xlcPersonalityV0 = &__xlcxx_personality_v0;
+        // Resolve __xlcxx_personality_v0 using dlopen().
+        const char libcxxabi[] = "libc++abi.a(libc++abi.so.1)";
```
[the remark this answers is not in the document]
Comment (xingxue-ibm) [reply libunwind/src/UnwindCursor.hpp]: Changed as suggested, thanks!

**Comment by xingxue-ibm** [comment]
> I think a small test case could be added with a `int main(void) {}` C program linked with -`brtl` and `LDR_PRELOAD`/`LDR_PRELOAD64` set to `libunwind.a(libunwind.so.1)`.

Added a test case `aix_runtime_link.pass.cpp` although the test case needs to be in C. The reason for the `.cpp` suffix is the libunwind test configuration does not recognize `.c` as a suffix for test cases. Also, `%{cxx}` is used to compile. However, the test configuration uses `-nostdlib++ -L %{lib} -lunwind -ldl -Wl,-bbigtoc` as link flags so it does not link with `libc++abi` and the test case does fail without the change.

**Review by hubert-reinterpretcast — APPROVED** [review]
LGTM! Thanks!

**Comment by xingxue-ibm** [comment]
Since this is an AIX-specific issue and the fix has been approved by AIX compiler and runtime expert @hubert-reinterpretcast (thank you), I plan to land it tomorrow if there are no objections.

**Pull request merged**